### PR TITLE
Normalize enum value to ClassConstFetch

### LINKED ISSUE #930: `Enum` as a `Attribute` parameter
Hello!

I think I found bug. Lets consider following snippet:

```php
<?php

use PhpParser\BuilderFactory;
use PhpParser\Node\Name;
use PhpParser\PrettyPrinter\Standard;

enum HttpMethod: string
{
    case GET = 'GET';
}

#[Attribute(Attribute::TARGET_METHOD)]
readonly class HTTP
{
    public HttpMethod $httpMethod;

    public function __construct(
        HttpMethod|string $httpMethod,
        public string $path
    )
    {
        if (is_string($httpMethod)) {
            $httpMethod = HttpMethod::from($httpMethod);
        }
        $this->httpMethod = $httpMethod;
    }
}

interface Api
{
    #[HTTP(httpMethod: HttpMethod::GET, path: '/users')]
    public function getUsers();
}

$reflectionClass = new ReflectionClass(Api::class);

$builderFactory = new BuilderFactory();

$class = $builderFactory->class('ApiImpl')
    ->implement('\\' . Api::class);

$reflectionMethod = $reflectionClass->getMethod('getUsers');

$method = $builderFactory
    ->method($reflectionMethod->getName())
    ->makePublic();

$reflectionAttributes = $reflectionMethod->getAttributes();
foreach ($reflectionAttributes as $reflectionAttribute) {
    $name = new Name($reflectionAttribute->getName());
    $attribute = $builderFactory->attribute($name, $reflectionAttribute->getArguments());

    $method->addAttribute($attribute);
}

$class->addStmt($method->getNode());

$proxyNamespaceBuilder = $builderFactory
    ->namespace('Namespace')
    ->addStmt($class);

$standard = new Standard();
echo $standard->prettyPrint([$proxyNamespaceBuilder->getNode()]);
```

When I try to create a proxy class with method which takes `enum` as a parameters I receive an exception:

```
Fatal error: Uncaught LogicException: Invalid value in /opt/project/vendor/nikic/php-parser/lib/PhpParser/BuilderHelpers.php on line 280

Call Stack:
    0.0004     401096   1. {main}() /opt/scratches/scratch_37.php:0
    0.0052    1342792   2. PhpParser\BuilderFactory->attribute($name = class PhpParser\Node\Name { protected $attributes = []; public $parts = [0 => 'HTTP'] }, $args = ['httpMethod' => enum HttpMethod::GET('GET'), 'path' => '/users']) /opt/scratches/scratch_37.php:54
    0.0052    1346616   3. PhpParser\BuilderFactory->args($args = ['httpMethod' => enum HttpMethod::GET('GET'), 'path' => '/users']) /opt/project/vendor/nikic/php-parser/lib/PhpParser/BuilderFactory.php:26
    0.0053    1351336   4. PhpParser\BuilderHelpers::normalizeValue($value = enum HttpMethod::GET('GET')) /opt/project/vendor/nikic/php-parser/lib/PhpParser/BuilderFactory.php:252
```

When I added a following handling in `BuilderHelpers::normalizeValue()` code works OK.

```php
public static function normalizeValue($value) : Expr {
    if ($value instanceof Node\Expr) {
        return $value;
    }

    if (is_null($value)) {
        return new Expr\ConstFetch(
            new Name('null')
        );
    }

    if (is_bool($value)) {
        return new Expr\ConstFetch(
            new Name($value ? 'true' : 'false')
        );
    }

    if (is_int($value)) {
        return new Scalar\LNumber($value);
    }

    if (is_float($value)) {
        return new Scalar\DNumber($value);
    }

    if (is_string($value)) {
        return new Scalar\String_($value);
    }

    if (is_array($value)) {
        $items = [];
        $lastKey = -1;
        foreach ($value as $itemKey => $itemValue) {
            // for consecutive, numeric keys don't generate keys
            if (null !== $lastKey && ++$lastKey === $itemKey) {
                $items[] = new Expr\ArrayItem(
                    self::normalizeValue($itemValue)
                );
            } else {
                $lastKey = null;
                $items[] = new Expr\ArrayItem(
                    self::normalizeValue($itemValue),
                    self::normalizeValue($itemKey)
                );
            }
        }

        return new Expr\Array_($items);
    }

    // !!! HERE !!!
    if (is_object($value)) {
        return new Expr\ClassConstFetch(new Name($value::class), $value->name);
    }

    throw new \LogicException('Invalid value');
}
```

What do you think about this fix? It's ok? Or it should be handled in the another way?

### PR DESCRIPTION
**Pull request by ruudk**
Fixes #930

### CODE REVIEW

File: test/PhpParser/BuilderHelpersPHP81Test.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+use PhpParser\Node\Name\FullyQualified;
+
+class BuilderHelpersPHP81Test extends \PHPUnit\Framework\TestCase
+{
+    public function testNormalizeValueEnum() {
+        if (\PHP_VERSION_ID <= 80100) {
+            $this->markTestSkipped('Enums are supported since PHP 8.1');
+        }
+
+        include __DIR__ . '/../code/Suit.php';
```
Comment (ruudk) [inline test/PhpParser/BuilderHelpersPHP81Test.php:15]: @nikic Is this the correct way to handle this?

**Comment by vitek-rostislav** [comment]
Hi, just encountered the problem described in https://github.com/nikic/PHP-Parser/issues/930. I can confirm the solution proposed   here fixes the issue. What is the status of this PR, please? It has been pending for a while now

**Comment by ruudk** [comment]
@nikic could you have a look please 🙏🏻

**Comment by adrianpaiva1** [comment]
@ruudk @nikic any update on this? I'm having the same issue.

**Comment by ruudk** [comment]
Waiting for a merge 😁

File: test/PhpParser/BuilderHelpersPHP81Test.php (hunk 1 above)
Comment (nikic) [reply test/PhpParser/BuilderHelpersPHP81Test.php:16]: This seems reasonable (alternative would be eval). But this shouldn't be inside the code/ directory.

File: lib/PhpParser/BuilderHelpers.php [hunk 2]
```diff
@@ -215,7 +216,7 @@ public static function normalizeType($type) {
      * Normalizes a value: Converts nulls, booleans, integers,
      * floats, strings and arrays into their respective nodes
      *
-     * @param Node\Expr|bool|null|int|float|string|array $value The value to normalize
+     * @param Node\Expr|bool|null|int|float|string|array|\UnitEnum $value The value to normalize
```
Comment (nikic) [inline lib/PhpParser/BuilderHelpers.php:219]: This doc comment change also needs to be propagated to most users of normalizeValue().

File: test/PhpParser/BuilderHelpersPHP81Test.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?php declare(strict_types=1);
+
+namespace PhpParser;
+
+use PhpParser\Node\Expr;
+use PhpParser\Node\Identifier;
+use PhpParser\Node\Name\FullyQualified;
+
+class BuilderHelpersPHP81Test extends \PHPUnit\Framework\TestCase
```
Comment (nikic) [inline test/PhpParser/BuilderHelpersPHP81Test.php:9]: Can merge this into the main BuilderHelpersTest, as the PHP version condition is inside the method anyway?

File: test/PhpParser/BuilderHelpersPHP81Test.php (hunk 1 above)
Comment (ruudk) [reply test/PhpParser/BuilderHelpersPHP81Test.php]: Created an `fixtures` directory and moved it there + rebased the PR.

File: lib/PhpParser/BuilderHelpers.php (hunk 2 above)
Comment (ruudk) [reply lib/PhpParser/BuilderHelpers.php:219]: I checked all usages, but I only added it to `\PhpParser\BuilderFactory::val`

File: test/PhpParser/BuilderHelpersPHP81Test.php (hunk 3 above)
Comment (ruudk) [reply test/PhpParser/BuilderHelpersPHP81Test.php]: Done!

File: lib/PhpParser/BuilderHelpers.php (hunk 2 above)
Comment (nikic) [reply lib/PhpParser/BuilderHelpers.php:219]: Why only that one? Enums are valid values for property/param/const defaults.

Comment (ruudk) [reply lib/PhpParser/BuilderHelpers.php:219]: Those mentioned are typed as `mixed`. 

But I found 2 more in `\PhpParser\Builder\ClassConst`: `__construct` and `addConst`.

Comment (ruudk) [reply lib/PhpParser/BuilderHelpers.php:219]: Done

**Review by nikic — APPROVED** [review]

**Pull request merged**